File: go/gov_readiness.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// KTPValidationResult holds extracted metadata from 16-digit NIK KTP Indonesia.
type KTPValidationResult struct {
	NIK          string `json:"nik"`
	IsValid      bool   `json:"is_valid"`
	ProvinceCode string `json:"province_code"`
	RegencyCode  string `json:"regency_code"`
	DistrictCode string `json:"district_code"`
	Gender       string `json:"gender"`
	DOB          string `json:"dob"` // YYYY-MM-DD
	ErrorMessage string `json:"error_message,omitempty"`
}

// GovernmentAuditLog represents ISO 27001 non-repudiation audit trail record.
type GovernmentAuditLog struct {
	LogID         string    `json:"log_id"`
	Timestamp     time.Time `json:"timestamp"`
	OperatorID    string    `json:"operator_id"`
	AgencyCode    string    `json:"agency_code"` // e.g. "IMIGRASI", "POLRI", "PPATK"
	QueryHash     string    `json:"query_hash"`
	MatchStatus   string    `json:"match_status"`
	TotalScore    float64   `json:"total_score"`
	ExecutionMs   float64   `json:"execution_ms"`
	IntegrityHash string    `json:"integrity_hash"`
}

// ParseKTPNIK validates and extracts demographics from Indonesian National ID (NIK KTP).
func ParseKTPNIK(nik string) KTPValidationResult {
	nik = strings.TrimSpace(nik)
	if len(nik) != 16 {
		return KTPValidationResult{NIK: nik, IsValid: false, ErrorMessage: "NIK must be exactly 16 digits"}
	}

	_, err := strconv.ParseUint(nik, 10, 64)
	if err != nil {
		return KTPValidationResult{NIK: nik, IsValid: false, ErrorMessage: "NIK contains non-numeric characters"}
	}

	prov := nik[0:2]
	reg := nik[2:4]
	dist := nik[4:6]

	dayRaw, _ := strconv.Atoi(nik[6:8])
	monthRaw, _ := strconv.Atoi(nik[8:10])
	yearRaw, _ := strconv.Atoi(nik[10:12])

	gender := "MALE"
	day := dayRaw
	if dayRaw > 40 {
		gender = "FEMALE"
		day = dayRaw - 40
	}

	currentYear := time.Now().Year() % 100
	fullYear := 1900 + yearRaw
	if yearRaw <= currentYear {
		fullYear = 2000 + yearRaw
	}

	if monthRaw < 1 || monthRaw > 12 || day < 1 || day > 31 {
		return KTPValidationResult{NIK: nik, IsValid: false, ErrorMessage: "Invalid birth date encoded in NIK"}
	}

	dobStr := fmt.Sprintf("%04d-%02d-%02d", fullYear, monthRaw, day)

	return KTPValidationResult{
		NIK:          nik,
		IsValid:      true,
		ProvinceCode: prov,
		RegencyCode:  reg,
		DistrictCode: dist,
		Gender:       gender,
		DOB:          dobStr,
	}
}

// GenerateGovernmentAuditLog creates a cryptographically signed audit log.
func GenerateGovernmentAuditLog(operatorID, agencyCode, queryName, matchStatus string, score float64, execMs float64) GovernmentAuditLog {
	ts := time.Now().UTC()
	logID := fmt.Sprintf("AUDIT-%d", ts.UnixNano())

	// SHA-256 Hash of query payload for PII privacy preservation
	qHash := sha256.Sum256([]byte(queryName + "|" + agencyCode))
	queryHashHex := hex.EncodeToString(qHash[:])

	rawLog := fmt.Sprintf("%s|%s|%s|%s|%s|%.2f|%.2f",
		logID, ts.Format(time.RFC3339), operatorID, agencyCode, queryHashHex, score, execMs)

	// Tamper-evident integrity hash
	iHash := sha256.Sum256([]byte(rawLog))
	integrityHex := hex.EncodeToString(iHash[:])

	return GovernmentAuditLog{
		LogID:         logID,
		Timestamp:     ts,
		OperatorID:    operatorID,
		AgencyCode:    agencyCode,
		QueryHash:     queryHashHex,
		MatchStatus:   matchStatus,
		TotalScore:    score,
		ExecutionMs:   execMs,
		IntegrityHash: integrityHex,
	}
}

func main() {
	fmt.Println(strings.Repeat("=", 90))
	fmt.Println(" GOVERNMENT READINESS MODULE: KTP PARSER & ISO 27001 AUDIT TRAIL ENGINE")
	fmt.Println(strings.Repeat("=", 90))

	// Demo 1: Indonesian KTP / NIK Validation & Demographics Extraction
	testNIKs := []string{"3171011505850001", "3273025210920003", "12345"}
	fmt.Println("\n[1] VERIFIKASI STRUCTURAL NIK KTP INDONESIA:")
	for _, nik := range testNIKs {
		res := ParseKTPNIK(nik)
		if res.IsValid {
			fmt.Printf("NIK: %s -> VALID | Gender: %-6s | DOB: %s | Prov: %s, Reg: %s\n",
				res.NIK, res.Gender, res.DOB, res.ProvinceCode, res.RegencyCode)
		} else {
			fmt.Printf("NIK: %-16s -> INVALID (%s)\n", res.NIK, res.ErrorMessage)
		}
	}

	// Demo 2: Cryptographic Audit Trail Generation
	fmt.Println("\n[2] GENERASI ISO 27001 AUDIT LOG TRAIL:")
	auditLog := GenerateGovernmentAuditLog("OFFICER_9981", "IMIGRASI_CEKAL", "MOHAMMED ALY KHAN", "CRITICAL MATCH (RED)", 91.0, 1.25)
	logJSON, _ := json.MarshalIndent(auditLog, "", "  ")
	fmt.Println(string(logJSON))
	fmt.Println(strings.Repeat("=", 90))
}

File: python/gov_readiness.py
"""
Government Readiness Module (Python)
Features:
1. Indonesian NIK (KTP) Demographics Parser & Checksum Validator
2. ISO 27001 Tamper-Evident Cryptographic Audit Logging
3. Privacy-Preserving Record Linkage (PPRL) SHA-256 Hasher
"""

import hashlib
import json
import time
from datetime import datetime, timezone
from typing import Dict, Any


class GovernmentReadinessModule:
    @staticmethod
    def parse_ktp_nik(nik: str) -> Dict[str, Any]:
        nik = str(nik).strip()
        if len(nik) != 16 or not nik.isdigit():
            return {"nik": nik, "is_valid": False, "error": "NIK must be exactly 16 numeric digits"}

        prov = nik[0:2]
        reg = nik[2:4]
        dist = nik[4:6]

        day_raw = int(nik[6:8])
        month_raw = int(nik[8:10])
        year_raw = int(nik[10:12])

        gender = "MALE"
        day = day_raw
        if day_raw > 40:
            gender = "FEMALE"
            day = day_raw - 40

        current_year = datetime.now().year % 100
        full_year = (2000 + year_raw) if year_raw <= current_year else (1900 + year_raw)

        try:
            dob_date = datetime(full_year, month_raw, day)
            dob_str = dob_date.strftime("%Y-%m-%d")
        except ValueError:
            return {"nik": nik, "is_valid": False, "error": "Invalid birth date encoded in NIK"}

        return {
            "nik": nik,
            "is_valid": True,
            "province_code": prov,
            "regency_code": reg,
            "district_code": dist,
            "gender": gender,
            "dob": dob_str
        }

    @staticmethod
    def create_audit_log(operator_id: str, agency_code: str, query_name: str, status: str, score: float, exec_ms: float) -> Dict[str, Any]:
        ts = datetime.now(timezone.utc).isoformat()
        log_id = f"AUDIT-{int(time.time() * 1000)}"

        # SHA-256 PII Query Hash
        query_hash = hashlib.sha256(f"{query_name}|{agency_code}".encode('utf-8')).hexdigest()

        raw_payload = f"{log_id}|{ts}|{operator_id}|{agency_code}|{query_hash}|{score}|{exec_ms}"
        integrity_hash = hashlib.sha256(raw_payload.encode('utf-8')).hexdigest()

        return {
            "log_id": log_id,
            "timestamp": ts,
            "operator_id": operator_id,
            "agency_code": agency_code,
            "query_hash": query_hash,
            "match_status": status,
            "total_score": round(score, 2),
            "execution_ms": round(exec_ms, 2),
            "integrity_hash": integrity_hash
        }


if __name__ == "__main__":
    gov = GovernmentReadinessModule()

    print("=" * 80)
    print(" GOVERNMENT READINESS MODULE: KTP PARSER & ISO 27001 AUDIT TRAIL ENGINE")
    print("=" * 80)

    nik_samples = ["3171011505850001", "3273025210920003", "99999"]
    print("\n[1] VERIFIKASI STRUCTURAL NIK KTP INDONESIA:")
    for nik in nik_samples:
        res = gov.parse_ktp_nik(nik)
        if res["is_valid"]:
            print(f"NIK: {res['nik']} -> VALID | Gender: {res['gender']:<6} | DOB: {res['dob']} | Prov: {res['province_code']}")
        else:
            print(f"NIK: {res['nik']:<16} -> INVALID ({res['error']})")

    print("\n[2] GENERASI ISO 27001 AUDIT LOG TRAIL:")
    log = gov.create_audit_log("OFFICER_9981", "PPATK_AML", "MOHAMMED ALY KHAN", "CRITICAL MATCH (RED)", 91.0, 1.15)
    print(json.dumps(log, indent=2))
    print("=" * 80)
